core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1348,24 +1223,61 @@ public void prepareCommit() throws IOException {
 
     @Override
     public void commit() throws IOException {
-        ImmutableTranslogReader toClose = null;
         try (ReleasableLock lock = writeLock.acquire()) {
             ensureOpen();
-            if (currentCommittingTranslog == null) {
+            if (currentCommittingGeneration == NOT_SET_GENERATION) {
                 prepareCommit();
             }
+            assert currentCommittingGeneration != NOT_SET_GENERATION;
             lastCommittedTranslogFileGeneration = current.getGeneration(); // this is important - otherwise old files will not be cleaned up
-            if (recoveredTranslogs.isEmpty() == false) {
-                IOUtils.close(recoveredTranslogs);
-                recoveredTranslogs.clear();
+            currentCommittingGeneration = NOT_SET_GENERATION;
+            trimUnreferencedReaders();
+        }
+    }
+
+    void trimUnreferencedReaders() {
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            if (closed.get()) {
+                // we're shutdown potentially on some tragic event - don't delete anything
+                return;
+            }
+            long minReferencedGen = outstandingViews.stream().mapToLong(View::minTranslogGeneration).min().orElse(Long.MAX_VALUE);
+            minReferencedGen = Math.min(lastCommittedTranslogFileGeneration, minReferencedGen);
+            while (readers.isEmpty() == false) {
+                ImmutableTranslogReader reader = readers.get(0);
+                if (reader.getGeneration() >= minReferencedGen) {
+                    return;
+                }
+                readers.remove(0);
+                Path translogPath = reader.path();
+                logger.trace("delete translog file - not referenced and not current anymore {}", translogPath);
+                IOUtils.closeWhileHandlingException(reader);
+                IOUtils.deleteFilesIgnoringExceptions(translogPath);
+                IOUtils.deleteFilesIgnoringExceptions(translogPath.resolveSibling(getCommitCheckpointFileName(reader.getGeneration())));
+            }
+        }
+    }
+
+    void closeFilesIfNoPendingViews() {
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            if (closed.get() == false) {
+                // we are not closed yet.
+                return;
+            }
+            if (outstandingViews.isEmpty()) {
+                logger.trace("closing files. translog is closed and there are no pending views");
+                try {
+                    ArrayList<Closeable> toClose = new ArrayList<>(readers);
+                    toClose.add(current);
+                    IOUtils.close(toClose);
+                } catch (IOException e) {
+                    logger.error("unexpected error while closing files", e);
```

Comment:
It was the way of some interface before. Double checking now I think it's doable. Good catch.